packages/xstate-solid/src/fsm.ts

```diff
@@ -0,0 +1,73 @@
+import type {
+  MachineImplementationsFrom,
+  ServiceFrom,
+  StateFrom,
+  StateMachine
+} from '@xstate/fsm';
+import { createMachine, interpret } from '@xstate/fsm';
+
+import { createStore, reconcile } from 'solid-js/store';
+import type { Accessor } from 'solid-js';
+import { createEffect, createMemo, on, onCleanup } from 'solid-js';
+import { deepClone } from './utils';
+
+type UseFSMReturnTuple<TService extends StateMachine.AnyService> = [
+  StateFrom<TService>,
+  TService['send'],
+  TService
+];
+export function useMachine<TMachine extends StateMachine.AnyMachine>(
+  machine: TMachine,
+  options?: MachineImplementationsFrom<TMachine>
+): UseFSMReturnTuple<ServiceFrom<TMachine>> {
+  const resolvedMachine = createMachine(
+    machine.config,
+    options ? options : (machine as any)._options
+  );
+
+  const service = interpret(resolvedMachine).start();
+
+  return useService(service) as UseFSMReturnTuple<ServiceFrom<TMachine>>;
+}
+
+export function useService<TService extends StateMachine.AnyService>(
+  service: TService | Accessor<TService>
+): UseFSMReturnTuple<TService> {
+  const serviceMemo = createMemo(() =>
+    typeof service === 'function' ? service() : service
+  );
+
+  const getClonedState = () =>
+    deepClone(serviceMemo().state) as StateFrom<TService>;
+
+  const [state, setState] = createStore<StateFrom<TService>>({
+    ...(getClonedState() as object),
+    matches(...args: Parameters<StateFrom<TService>['matches']>) {
+      // tslint:disable-next-line:no-unused-expression
+      (state as StateFrom<any>).value; // sets state.value to be tracked by the store
+      return serviceMemo().state.matches(args[0] as never);
+    }
+  } as StateFrom<TService>);
+
+  // Track if a new service is passed in, only update once per service
+  createEffect(
```

Comment:
I have no idea what's the expectation in the Solid world but in React this kinda should be executed asap, so perhaps this should use a `createRenderEffect`?
```suggestion
  createRenderEffect(
```

The reason for this in the React world is that when the "state container" changes during an update we want the next committed screen to be based on that new state container. OTOH `useEffect` defers the `setState` to the end of the commit phase - so with a regular effect, we get 2 renders: one to update the state container and one to update the actual value that it provides.

I won't be dying on this hill - so just let me know what do you think is best here.

Reply:
Note that things get tricky with something like `useRenderEffect` here because we'd have to make sure that the old `service` is ignored after receiving the new `service` (or that its updates are ignored). Essentially we want to prevent this situation:
```
useRenderEffect -> seState
old subscribe's callback -> setState
old useEffect -> cleanup
new useEffect -> setup
```

We can't accept the update from the stale~ subscription. I'm actually not super sure if React can currently avoid extra renders here, but at least their render won't return a stale data in such a case because they re-execute hooks and the very latest value is always returned from there:
https://github.com/facebook/react/blob/b2748907c3cffb226447417957f2608a7c60c27d/packages/react-reconciler/src/ReactFiberHooks.new.js#L1569

The effect-syncing, subscribing, and updating the store is a weird dance there though and I'm not sure if I can quickly deduce all of the implications from that code. It's somewhat easier to reason about their shim but it still involved that delicate dance:
https://unpkg.com/browse/use-sync-external-store@1.2.0/cjs/use-sync-external-store-shim.development.js

Reply:
👋